api/base/filters.py

```diff
@@ -61,6 +63,78 @@ def filter_queryset(self, request, queryset, view):
         return queryset
 
 
+class MultiValueCharFilter(django_filters.BaseInFilter, django_filters.filters.CharFilter):
+
+    def filter(self, qs, value):
+        q = Q()
+        values = value or []
+        for value in values:
+            name = self.name
+            q |= Q(**{'{}__icontains'.format(name): value})
+        qs = qs.filter(q)
+
+        return qs
+
+
+class JSONAPIFilterSet(django_filters.FilterSet):
+
+    QUERY_PATTERN = re.compile(r'^filter\[(?P<fields>((?:,*\s*\w+)*))\](\[(?P<op>\w+)\])?$')
+    FILTER_FIELDS = re.compile(r'(?:,*\s*(\w+)+)')
+
+    or_fields = {}
+
+    def __init__(self, data=None, *args, **kwargs):
+        self.or_fields = {}
+        field_names = []
+        if data:
+            new_data = {}
+            for key, value in data.iteritems():
+                match = self.QUERY_PATTERN.match(key)
+                if match:
+                    match_dict = match.groupdict()
+                    fields = match_dict['fields']
+                    field_names = re.findall(self.FILTER_FIELDS, fields.strip())
+                    values = value.split(',')
+                    if len(field_names) > 1:
+                        self.or_fields[frozenset(field_names)] = values
+                    for field in field_names:
+                        new_data.update({field: value})
+            data = new_data
+        super(JSONAPIFilterSet, self).__init__(data=data, *args, **kwargs)
+
+        for field in field_names:
+            if field not in self.form.fields.keys():
+                raise InvalidFilterFieldError(parameter='filter', value=field)
+
+    @property
+    def qs(self, *args, **kwargs):
+        self.form.is_valid()
+        or_keys = reduce(operator.or_, self.or_fields.keys(), set())
+        for key in or_keys:
+            self.form.cleaned_data.pop(key, None)
```

Comment:
Rather than removing these keys from `form.cleaned_data`, would it make sense to prevent these fields from going through the form validation in the first place, i.e. preventing all the or_fields from going into `self.form.fields`?